## C12 - P3

## (Computer Problem 12.2.3ab) Apply the shifted QR algorithm directly to find all the eigenvalues of the following matrices.

## (a) $$A = \begin{equation*} \begin{bmatrix} -1&1&3\\ 3&3&-2 \\ -5&2&7 \end{bmatrix} \end{equation*} $$

In [1]:
import numpy as np
import shiftedqr as myqr

aa = np.array([[-1.,1.,3.],[3.,3.,-2.],[-5.,2.,7.]])
lam = myqr.shiftedqr(aa)
print('\neigenvals of aa = ', lam)

print('\nnumpy eigenvals are', np.linalg.eig(aa)[0])


eigenvals of aa =  [ 2.99997994 +0.00000000e+00j  3.00001003 -1.73716380e-05j
  3.00001003 +1.73716380e-05j]

numpy eigenvals are [ 3.00001088 +1.88367701e-05j  3.00001088 -1.88367701e-05j
  2.99997825 +0.00000000e+00j]


## The computed eigenvalues computed are , $[2.99997994, 3.00001003, 3.00001003]$. These are correct to $4$ decimal places.


## (b) $$A = \begin{equation*} \begin{bmatrix} 7&-33&-15\\ 2&26&7 \\-4&-50&-13 \end{bmatrix} \end{equation*} $$

In [2]:
bb = np.array([[7.,-33.,-15.],[2.,26.,7.],[-4.,-50.,-13.]])
lam = myqr.shiftedqr(bb)
print('\nno-hess, eigenvals bb = ', lam)

print('\nnumpy eigenvals are', np.linalg.eig(bb)[0])


no-hess, eigenvals bb =  [10.+0.j  9.+0.j  1.+0.j]

numpy eigenvals are [10.  9.  1.]


## The computed eigenvalues computed are , $[10.0, 9.0, 1.0]$. These are the correct eigenvalues.